kic_user.c

```diff
@@ -0,0 +1,148 @@
+/*
+ * userspace applicaton to control and comunicate with associated kernel module
+ * can be accessed through command line arguments "count", "reset" and "time"
+ * or a simple text menu when no arguments are provided
+ */
+
+#include<stdio.h>
+#include<stdlib.h>
+#include<fcntl.h>
+#include<sys/ioctl.h>
+#include<time.h>
+#include<string.h>
+
+#include"keyboard_interrupt_counter.h"
+
+// prototypes
+int retrieve_count(int);
+/*
+ * prints out current interrupt count provided by kernel,
+ * returns 0 on success or error code on failure
+ */
+
+int reset_count(int);
+/*
+ * tasks the cernel with reseting interrupt counter,
+ * returns 0 on success or error code on failure
+ */
+
+int retrieve_time(int);
+/*
+ * prints out date and time of last counter reset,
+ * returns 0 on success or error code on failure
+ */
+
+void menu(int);
+/*
+ * prints out options provided to the user and returns requested data by
+ * invoking other fuctions necessary
+ */
+
+
+int main(int argc, char *argv[])
+{
+	char choice[10];
+	int file_descriptor, ret = 0;
+	// accesing the device file
+	file_descriptor = open(KIC_DEVICE_PATH, O_RDWR);
+	if (file_descriptor < 0) {
+		printf("Device:%s inaccesible or missing. Error Code: %d\n", KIC_DEVICE_PATH, file_descriptor);
+		exit(EXIT_FAILURE);
+	}
+
+	if(argc == 1)
+		menu(file_descriptor);
+	else
+		if (!strcmp(argv[1], "count")) {
+			retrieve_count(file_descriptor);
+		} else if (!strcmp(argv[1], "reset")) {
+			reset_count(file_descriptor);
+		} else if (!strcmp(argv[1], "time")) {
+			retrieve_time(file_descriptor);
+		} else
+			printf("Operation unsupported. Check spelling.\n");
+	return 0;
+}
+
+
+int retrieve_count(int file_descriptor)
+{
+	int result = 0;
+	unsigned int data_carry;
+	result = ioctl(file_descriptor, KIC_GET_COUNT_QUERY, &data_carry);
+
+	if (result < 0)
+		printf("Retrieval Operation failed. Error Code: %d\n", result);
+	else
+		printf("Current keyboard interrupt count: %u\n", data_carry);
+
+	return result;
+}
+
+int reset_count(int file_descriptor)
+{
+	int result = 0;
+	result = ioctl(file_descriptor, KIC_RESET_QUERY);
```

Comment:
empty line before this line
generally we do a section of variables declarations, empty line, the rest of code

Reply:
done